distribution/cachedir/manager.go

```diff
@@ -0,0 +1,214 @@
+// very particular way of managing the cache. It was designed to be used by a
+// caller that manages the lifetime of the directories, but delegates the
+// mechanics of how to store and locate the directories to the Manager
+type Manager interface {
+	// GetDir creates a new cache dir or returns an existing one if one already
+	// exists for this key from a previous download. It also keeps track of how
+	// many references were made to the cache dir during
+	// ReleaseDir or DeleteDir have to be called exactly once per GetDir call
+	// and GetDir shouldn't be called again until one or the other is called.
+	GetDir(key string) string
+	// ReleaseDir lets the Manager know that the dir is no longer in use but
+	// should be kept around in case we need it again.
+	ReleaseDir(key string)
+	// DeleteDir lets the manager know that the dir is no longer used and can be
+	// deleted. If the delete operation fails and returns an error, the key is
+	// still considered released.
+	DeleteDir(key string) error
+	// CollectGarbage deletes all dirs that are not currently in use (ones that
+	// don't have a GetDir without a matching ReleaseDir/DeleteDir call) and
+	// returns the amount of space saved in bytes.
+	CollectGarbage() (uint64, error)
+	// Usage returns information about how much space the download cache is
+	// using. If any directories or files are unreadable, we just log the
+	// errors and don't count the files.
+	Usage() types.DownloadCacheUsage
+}
+
+type manager struct {
+	cacheRoot string
+	mu        sync.Mutex
+	inUse     map[digest.Digest]struct{}
+}
+
+// NewManager creates a new cachedir manager
+func NewManager(path string) Manager {
+	return &manager{
+		cacheRoot: path,
+		inUse:     map[digest.Digest]struct{}{},
+	}
+}
+
+func (m *manager) cachePathForKeyHash(keyHash digest.Digest) string {
+	return filepath.Join(m.cacheRoot, keyHash.Algorithm().String(), keyHash.Hex())
+}
+
+func dirSize(path string) (uint64, error) {
+	var size uint64
+	err := filepath.Walk(path, func(_ string, info os.FileInfo, err error) error {
+		if info != nil && !info.IsDir() {
+			size += uint64(info.Size())
+		}
+		return err
+	})
+	return size, err
+}
+
+func (m *manager) sizeEstimateByKeyHash(keyHash digest.Digest) uint64 {
+	dir := m.cachePathForKeyHash(keyHash)
+	size, err := dirSize(dir)
+	if err != nil {
+		logrus.WithError(err).WithField("cachedir", dir).Warn("failed to determine size of directory")
+		return 0
+	}
+	return size
+}
+
+func (m *manager) cleanUpByKeyHash(keyHash digest.Digest) error {
+	return os.RemoveAll(m.cachePathForKeyHash(keyHash))
+}
+
+func (m *manager) GetDir(key string) string {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	hashedKey := digest.FromString(key)
+	logrus.WithField("key", key).WithField("hashedKey", hashedKey).Debug("cachedir manager get")
+	if _, ok := m.inUse[hashedKey]; ok {
+		panic(fmt.Errorf("tried to get a cachedir key (%s) twice without releasing/deleting", key))
```

Comment:
I am relying on the caller to do the deduplication. If I didn't do that, I would have to put deduplication logic here which would never be used, so I chose to encode my assumption about the way that the cachedir manager should be used in this panic (as well as in the comments). This is a panic because it should never happen if there is no mistake in the code. If we change the assumption that the caller will take care of deduplicating calls to the cachedir manager, then we should change this behaviour, but right now I think the panic expresses my intent more accurately than an erro would. If you guys still disagree that it should be a panic, I'll change it.